subsys/power/device_pm.c

```diff
@@ -0,0 +1,158 @@
+/*
```

Comment:
This does not look like the FSM we agreed on during our call.
Do you need some kind of design proposal?

Reply:
Before I implement the FSM I wanted some clarity on the state transitions.

Consider the following scenario.
1. device idle requested suspend
2. pm_state change from ACTIVE -> SUSPENDING
3. driver tries to suspend the device but fails to do so
4. driver calls the callback with an error return code.
In this case, what should we set the device state to?

Also if the pm_state is in SUSPENDING and new request comes for SUSPEND what we should do? should we call the device_set_power_state() with new SUSPEND requests for return?

what action we should for new requests if the current case is SUSPENDING or RESUMING? we can not skip the new requests(ex: SUSPEND) even if it is the same as the previous (ex: SUSPENDING) request because the previous request might fail. It would be better to rely only on the actual state for queuing the requests.

This commit now uses the actual device PM states as pm_states so we don't need any translation of FSM to device states in the device_pm code. I feel this simplifies the whole logic.

Reply:
@pizi-nordic Please have look at my query above. Thanks.

Reply:
You cannot use actual device state because it not covers transition state.
Here is algorithm I have in my mind. I hope it explain all the cases you described.
As you see, the algorithm is trying to reduce number of calls to device_set_power_state().

1. Assumptions:
    - Device could change power state only when it is idle. There is no "forcing" power state, as this might render the system non-operational (for example shutting down SPI bus during write to SPI flash will corrupt data). Failure of setting power mode is a failure of device - you cannot longer rely on it.
    - Suspending and resuming operations cannot be terminated until complete.

2. Variables:
    - FSM state: (ACTIVE, SUSPENDED, SUSPENDING, RESUMING)
    - Target state: (ACTIVE, SUSPENDED).

3. Pseudocode:
```
device_suspend() {
	if (target_state == SUSPENDED) {
		/* Device will eventually reach the SUSPSENDED state */
		return;
	}

	target_state = SUSPENDED;
	trigger_fsm();
}

device_resume() {
	if (target_state == ACTIVE) {
		/* Device will eventually reach the ACTIVE state */
		return;
	}

	target_state = ACTIVE;
	trigger_fsm();
}

trigger_fsm() {
	run_through_workqueue(fsm);
}

/* As device_set_power_state() is asynchronous, this will no block workqueue */
fsm() {
	switch (fsm_state) {
		case SUSPENDING:
		case RESUMING:
			/* We have to wait */
			break;

		case ACTIVE:
			if (target_state == SUSPENDED) {
				device_set_power_state(SUSPENDED, trigger_fsm)
				fsm_state = SUSPENDING;
			}
			break;

		case SUSPENDED:
			if (target_state == ACTIVE) {
				device_set_power_state(ACTIVE, trigger_fsm)
				fsm_state = RESUMING;
			}
			break;
	}

	broadcast_state(fsm_state);
}

device_wait_suspended() {
	while (recv_broadcasted_state() != SUSPENDED) { sleep() }
}

device_wait_active() {
	while (recv_broadcasted_state() != ACTIVE) { sleep() }
}

device_suspend_sync() {
	device_suspend();
	device_wait_suspended();
}

device_resume_sync() {
	device_resume();
	device_wait_active();
}
```

This could be extended to cancel workqueue job if is is not longer needed, such shortcut must be carefully checked for race conditions. As you see multiple calls to suspend/resume are handled and the last request is valid for us.

Reply:
I will look into it and get back to you.

Reply:
>    * FSM state: (ACTIVE, SUSPENDED, SUSPENDING, RESUMING)
>    * Target state: (ACTIVE, SUSPENDED).

> device_suspend() {
> 	if (target_state == SUSPENDED) {
> 		/* Device will eventually reach the SUSPSENDED state */
> 		return;
> 	}
> 
> 	target_state = SUSPENDED;
> 	trigger_fsm();
> }
 
> device_resume() {
> 	if (target_state == ACTIVE) {
> 		/* Device will eventually reach the ACTIVE state */
> 		return;
> 	}
> 
> 	target_state = ACTIVE;
> 	trigger_fsm();
> }
> trigger_fsm() {
> 	run_through_workqueue(fsm);
> }
Are you suggesting we should have workquque in the device_pm.c? If so, do we need a saperate thread or we should submit the work on system thread?

> fsm() {
> 	switch (fsm_state) {
> 		case SUSPENDING:
> 		case RESUMING:
> 			/* We have to wait */
> 			break;
> 
> 		case ACTIVE:
> 			if (target_state == SUSPENDED) {
> 				device_set_power_state(SUSPENDED, trigger_fsm)
> 				fsm_state = SUSPENDING;
> 			}
> 			break;
> 
> 		case SUSPENDED:
> 			if (target_state == ACTIVE) {
> 				device_set_power_state(ACTIVE, trigger_fsm)
> 				fsm_state = RESUMING;
> 			}
> 			break;
> 	}
> 
> 	broadcast_state(fsm_state);
> }
is this the worked logic which maintains the states? also why broadcat fsm state?

Reply:
> is this the worked logic which maintains the states?

I am not fully getting this question. If you are asking which entity holds the device state: In the proposed model this is tracked by FSM.

> why broadcat fsm state?

Broadcasting the FSM state is needed for blocking version of the device_get/put. Whenever state is changed, a signal is sent. If FSM reached desired state, then you can exit from the blocking function.

Also, the application can track the device state and run some action if device reach desired state.

Reply:
addressed, at least when it comes to the system workqueue